OpenAI Responses API

kamillitman opened the pull request.

nefernata commented on src/llm_models/tools.rs:
```diff
@@ -0,0 +1,151 @@
+}
+
+impl LLMTools {
+    pub fn get_config_json(&self) -> Option<Value> {
```
Minor refactor: this impl should be moved to under the declaration of the LLMTools enum

kamillitman replied:
done.

nefernata commented on src/domain.rs:
```diff
@@ -375,3 +375,209 @@ pub struct DeepSeekAPICompletionsUsage {
+    pub max_output_tokens: Option<i32>,
+    pub model: String,
+    pub output: Vec<OpenAPIResponsesOutput>,
+    // pub previous_response_id: Option<String>,
```
A future PR may uncomment these to expose these fields in the response to the user.

nefernata commented on src/llm_models/openai.rs:
```diff
@@ -389,6 +495,148 @@ impl LLMModel for OpenAIModels {
+                    "max_output_tokens": max_tokens,
+                    "temperature": temperature,
+                    // If tools are provided we add them to the body
+                    "tools": tools.map(|tools_inner| tools_inner.iter().filter_map(LLMTools::get_config_json).collect::<Vec<Value>>()),
```
Filter out the reasoning tool here as it has its own way of getting attached to the message body

kamillitman replied:
done.

nefernata commented on src/llm_models/openai.rs:
```diff
@@ -389,6 +495,148 @@ impl LLMModel for OpenAIModels {
+                        {json_schema}"
+                    ),
+                    "instructions": base_instructions,
+                    // TODO: Check if this is correct
```
remove this TODO?

kamillitman replied:
done.

nefernata commented on src/llm_models/openai.rs:
```diff
@@ -389,6 +495,148 @@ impl LLMModel for OpenAIModels {
+                    "max_output_tokens": max_tokens,
+                    "temperature": temperature,
+                    // If tools are provided we add them to the body
+                    "tools": tools.map(|tools_inner| tools_inner.iter().filter_map(LLMTools::get_config_json).collect::<Vec<Value>>()),
```
Tools and temperature not allowed in the request body for the responses API

kamillitman replied:
done.

nefernata reviewed (changes requested):
Some changes required. Refer to comments

kamillitman commented on src/llm_models/openai.rs:
```diff
@@ -389,6 +495,148 @@ impl LLMModel for OpenAIModels {
+                        {json_schema}"
+                    ),
+                    "instructions": base_instructions,
+                    // TODO: Check if this is correct
```
done.

kamillitman commented on src/llm_models/openai.rs:
```diff
@@ -201,6 +201,11 @@ impl LLMModel for OpenAIModels {
                 | OpenAIModels::Custom { .. },
+            )
+            // o1 Pro is not supported in Completions API, we redirect to Responses API
+            | (
```
Instead of having a separate match branch for Completions + o1Pro, I added an OR (|) to a branch that returns the desired result already.

kamillitman commented:
> Some changes required. Refer to comments

I made all the requested changes. Additionally, I simplified the match statements. Previously we had separate branches for Completions + o1 Pro that were returning Responses endpoint, body and data formatting. But we already had branches that returned that for standard Responses route. So I eliminated the separate branches and added an OR statement. This helped to remove some of the code duplication.

nefernata reviewed (approved):
Changes appear to be made and the new tools additions look good. Approved
